### PR TITLE
Add psbt watch-only/cold-storage example

### LINKED ISSUE #892: Write an example psbt workflow
Write an example in the `examples` directory with the workflow from [BIP174](https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki#test-vectors). See the valid example that starts from the line in the above link.
```
The private keys in the tests below are derived from the following master private key:
```
- Use BIP32 wallet APIs to derive all priv/public keys
- Use the sighash APIs to compute the signature hashes
- Use all Psbt supported APIs
- Identify the API pain-points and suggest improvements.

### PR DESCRIPTION
**Pull request by tcharding**
Add an example PSBT workflow. The workflow we simulate is that of a setup using a watch-only online wallet (contains only public keys) and a cold-storage wallet (contains the private keys).

We create and update a PSBT using the watch-only wallet then pass the PSBT to the cold-storage wallet to sign.

Partially resolves #892 (more done in #935).

## Note

This PR includes a sub-module in `examples/psbt.rs` that implements ECDSA signing. This will hopefully eventually be merged into the main crate by way of https://github.com/rust-bitcoin/rust-bitcoin/pull/957.  We have three PRs that add examples/tests of PSBTs that need to do signing, in order to help us design a good AP in #957 I think it would be beneficial to complete and merge these three PRs.

1. This PR (PSBT ECDSA example)
2. PBST ECDSA integration test: https://github.com/rust-bitcoin/rust-bitcoin/pull/935
3. PSBT taproot example: https://github.com/rust-bitcoin/rust-bitcoin/pull/999 

Note to self, this will need a change to `test.sh` if #1079 merges.

### CODE REVIEW

File: examples/psbt.rs
Comment (DanGould) [inline examples/psbt.rs:262]: Though the updater has enough information to add `redeem_scripts` to `send` and `change` but there is no need. the unsigned_tx (PSBTv1 only) has the addresses which include the standard p2wpkh scripts already. With the input `redeem_script` set properly, the output of this program is a valid and correct bitcoin transaction as far as I can tell.

**Review by DanGould — CHANGES_REQUESTED** [review]

File: examples/psbt.rs
Comment (DanGould) [reply examples/psbt.rs:251]: I thought my suggestion was overkill. But for some reason

```rs
let redeem_script = Script::new_v0_p2wpkh(&wpkh);
```
returns
`Err(SignError::RedeemScriptMismatch)`.

Where did that error check come from?

File: examples/psbt.rs
Comment (DanGould) [inline examples/psbt.rs:301]: > The witness must consist of exactly 2 items (≤ 520 bytes each). The first one a signature, and the second one a public key.
```suggestion
        let mut script_witness: Witness = Witness::new();
        script_witness.push(&sigs[0].serialize());
        script_witness.push(self.input_xpub.to_pub().serialize()); // self.input_xpub.to_pub()
        psbt.inputs[0].final_script_witness = Some(script_witness);
```

File: examples/psbt.rs
Comment (DanGould) [inline examples/psbt.rs:290]: > There are two cases in which witness validation logic are triggered. Each case determines the location of the witness version byte and program, as well as the form of the scriptSig:
>    1. Triggered by a scriptPubKey that is exactly a push of a version byte, plus a push of a witness program. The scriptSig must be exactly empty or validation fails. ("native witness program")
```suggestion
```
I suppose you could set final_script_sig = None explicitly, too

File: examples/psbt.rs
Comment (DanGould) [reply examples/psbt.rs:251]: The error was due to the finalizer. I think it is working with my suggestions to the finalizer.

**Comment by tcharding** [comment]
Thanks @DanGould, you the man! I used all your suggestions and everything works now, I was able to broadcast the transaction with `bitcoind`. I added a `Co-authored-by` tag!

**Review by apoelstra — DISMISSED** [review]
ACK af148d182efcf37d79028e4a516bba75138fc273

Wow! Did not read the code *super* carefully but skimmed all of it. A very thorough example.

Can me merged before or after 0.28.0, it's completely independent of the release.

**Comment by tcharding** [comment]
I'd like to hold off on merging this until the bdk sighash license/upstream stuff is resolved.

**Comment by tcharding** [comment]
Converting this to draft until #953 gets sorted.

**Comment by tcharding** [comment]
Converting _back_ to draft because I've made changes to this over in https://github.com/rust-bitcoin/rust-bitcoin/pull/957 after review and cannot be bothered maintaining two branches.

**Comment by tcharding** [comment]
This PSBT example has been hanging around for months now. I've back and forthed, both mentally, and draft/undraft, a million times because of not knowing if its better to do #957 before or after this PR. Also, @DanGould has #935 open with similar confusion. Now @dunxen has #999 going to show taproot psbt signing.

Can we just merge these tests/examples and worry about cleaning them up to use #957 when/if that merges. My arguments are:
- Its a PITA having these examples/tests in limbo
- The other two devs are newish contributors and they did a fair bit of work on their PRs, it would be nice to recognize this and merge them instead of waiting on what is quite a curly PR (#957)
- We would only be merging into `tests`/ and `examples/` so this is not a risk to the library
- With three devs looking at this, at least in the short term, I don't think there is an undue a maintenance burden

Cheers

**Review by apoelstra — DISMISSED** [review]
ACK a7f2d210d213e60c758a801f053af273f46b24a0

**Comment by apoelstra** [comment]
Pretty cool! Yeah, let's just get this in. I could find some things to nit but it's not worth keeping this in limbo.

**Comment by tcharding** [comment]
sick!

**Comment by tcharding** [comment]
> I could find some things to nit but it's not worth keeping this in limbo.

Please do nit if you think its worth it and if it will help the other psbt example/test PRs. Otherwise I can come back and improve all three after they merge.

**Comment by tcharding** [comment]
Changes in force push:
- rebased
- Removed explicit scope required for Rust 1.29

**Review by apoelstra — APPROVED** [review]
ACK ee8d06dc474aca967a7cab8e0a56b2b308eb0453

**Comment by tcharding** [comment]
Changes in force push:
- Rebased to pick up `Sequence` changes
- Added a line to `test.sh` to run the example in all test runs (separate patch)

**Review by apoelstra — APPROVED** [review]
ACK cd2369b4f965267690878299b893fbaeded22e0c

**Comment by DanGould** [comment]
@sanket1729 this needs just one more approval

File: examples/ecdsa-psbt.rs
Comment (sanket1729) [inline examples/ecdsa-psbt.rs:266]: nit: You can push pk from the `key` in the partial_keys map here instead. It seems more natural to use that instead of deriving it again.

File: examples/ecdsa-psbt.rs
Comment (sanket1729) [inline examples/ecdsa-psbt.rs:170]: I would not call it `WatchOnly` because watch wallets do not create any transactions or assist in creating those. They just watch the blockchain for funds. This is more of an example of a software wallet that actively creates/finalizes transactions with an offline HW signer.

File: examples/ecdsa-psbt.rs
Comment (sanket1729) [inline examples/ecdsa-psbt.rs:176]: nit: I think we should save the derivation path for both `account_0` and `input_xpub` instead of global constants.

File: contrib/test.sh
Comment (sanket1729) [inline contrib/test.sh:65]: Not related to this issue, but I just noted that other examples are not being when "DO_NO_STD" is not set

**Review by sanket1729 — APPROVED** [review]
ACK cd2369b4f965267690878299b893fbaeded22e0c. The example is clean, you can address the comments in followups. 

A lot of the code would be cleaned up after #957

File: examples/ecdsa-psbt.rs
Comment (sanket1729) [reply examples/ecdsa-psbt.rs:170]: After some reading, I think WatchOnly is the correct term here

**Pull request merged**
